oximeter/db/src/oxql/ast/table_ops/filter.rs

```diff
@@ -152,82 +335,269 @@ impl Filter {
+                    };
+                    return Ok(Filter {
+                        negated: false,
+                        expr: FilterExpr::Compound(CompoundFilter {
+                            left: Box::new(left_),
+                            op: LogicalOp::Or,
+                            right: Box::new(right_),
+                        }),
+                    });
+                }
+
+                // This matches the rule:
+                //
+                // z && (x || y) -> (z && x) || (z && y)
+                if let (
+                    FilterExpr::Simple(z),
+                    LogicalOp::And,
+                    FilterExpr::Compound(CompoundFilter {
+                        left: x,
+                        op: LogicalOp::Or,
+                        right: y,
+                    }),
+                ) = (&left.expr, op, &right.expr)
+                {
+                    let left_ = Filter {
+                        negated: false,
+                        expr: FilterExpr::Compound(CompoundFilter {
+                            left: Box::new(z.to_filter()),
+                            op: LogicalOp::And,
+                            right: x.clone(),
+                        }),
+                    };
+                    let right_ = Filter {
+                        negated: false,
+                        expr: FilterExpr::Compound(CompoundFilter {
+                            left: Box::new(z.to_filter()),
+                            op: LogicalOp::And,
+                            right: y.clone(),
+                        }),
+                    };
+                    return Ok(Filter {
+                        negated: false,
+                        expr: FilterExpr::Compound(CompoundFilter {
+                            left: Box::new(left_),
+                            op: LogicalOp::Or,
+                            right: Box::new(right_),
+                        }),
+                    });
+                }
+
+                // Lastly, simplify an XOR to its logical equivalent, which is
+                // in DNF.
+                let out = Filter {
+                    negated: false,
+                    expr: FilterExpr::Compound(CompoundFilter {
+                        left: Box::new(left),
+                        op: *op,
+                        right: Box::new(right),
+                    }),
+                };
+                Ok(out.rewrite_xor_to_disjunction())
+            }
         }
     }
 }
 
-/// A more complicated expression as part of a filtering operation.
-///
-/// E.g., the `hostname == "bar" && id == "baz"` in the below.
-// NOTE: This should really be extended to a generic binary op expression.
+impl fmt::Display for FilterExpr {
+    fn fmt(&self, f: &mut fmt::Formatter<'_>) -> fmt::Result {
+        match self {
+            FilterExpr::Simple(inner) => write!(f, "{inner}"),
+            FilterExpr::Compound(inner) => write!(f, "{inner}"),
```

Comment:
If this is used for printing complex/nested expressions, would you want "({inner})" to visually group the compound?

Reply:
Hmm good question. The `Filter` type that contains a `FilterExpr` always wraps that in `()` already, so _think_ it should be OK? We always use the `Display` method on `Filter` when writing the expressions for the database, for example.